### PR TITLE
enhance: only buffer delete if it match insert has smaller timestamp

### PR DESCRIPTION
**Pull request by congqixia**
See also: #30121 #27675

This PR changes the delete buffering logic:
- Write buffer shall buffer insert first
- Then the delete messages shall be evaluated
  - Whether PK matches previous Bloom filter, which ts is always smaller
  - Whether PK matches insert data which has smaller timestamp
- Then the segment bloom filter is updates by the newly buffered pk rows

### CODE REVIEW

**Comment by yah01** [comment]
lgtm

File: internal/datanode/writebuffer/write_buffer.go [hunk 1]
```diff
@@ -325,51 +326,138 @@ func (wb *writeBufferBase) yieldBuffer(segmentID int64) (*storage.InsertData, *s
...
 
-		pkData, totalMemSize, err := segBuf.insertBuffer.Buffer(msgs, startPos, endPos)
-		if err != nil {
-			log.Warn("failed to buffer insert data", zap.Int64("segmentID", segmentID), zap.Error(err))
-			return nil, err
+// prepareInsert transfers InsertMsg into organized InsertData grouped by segmentID
+// also returns primary key field data
+func (wb *writeBufferBase) prepareInsert(insertMsgs []*msgstream.InsertMsg) ([]*inData, error) {
+	groups := lo.GroupBy(insertMsgs, func(msg *msgstream.InsertMsg) int64 { return msg.SegmentID })
+	segmentPartiton := lo.SliceToMap(insertMsgs, func(msg *msgstream.InsertMsg) (int64, int64) { return msg.GetSegmentID(), msg.GetPartitionID() })
```
Comment (XuanYang-cn) [inline internal/datanode/writebuffer/write_buffer.go:383]: typo "Partiton"

File: internal/datanode/writebuffer/write_buffer.go [hunk 2]
```diff
@@ -325,51 +326,138 @@ func (wb *writeBufferBase) yieldBuffer(segmentID int64) (*storage.InsertData, *s
...
+			}
+			if pkFieldData.RowNum() != data.GetRowNum() {
+				return nil, merr.WrapErrServiceInternal("pk column row num not match")
+			}
+
+			tsFieldData, err := storage.GetTimestampFromInsertData(data)
+			if err != nil {
+				return nil, err
+			}
+			if pkFieldData.RowNum() != data.GetRowNum() {
```
Comment (XuanYang-cn) [inline internal/datanode/writebuffer/write_buffer.go:412]: pkFieldData checked twice

File: internal/datanode/writebuffer/write_buffer.go (hunk 1 above)
Comment (congqixia) [reply internal/datanode/writebuffer/write_buffer.go]: fixed

File: internal/datanode/writebuffer/write_buffer.go (hunk 2 above)
Comment (congqixia) [reply internal/datanode/writebuffer/write_buffer.go]: fixed

**Pull request merged**